src/twisted/conch/ssh/filetransfer.py

```diff
@@ -289,10 +288,10 @@ def packet_OPENDIR(self, data):
 
 
     def _cbOpenDirectory(self, dirObj, requestId):
-        handle = ensureBytes((str(hash(dirObj))))
+        handle = networkString((str(hash(dirObj))))
         if handle in self.openDirs:
             raise KeyError("already opened this directory")
-        self.openDirs[handle] = [ensureBytes(dirObj), iter(dirObj)]
```

Comment:
This is an example of a bug created by `ensureBytes`.

Reply:
Many thanks for looking into this.

Do we have a test to cover this behavior and prevent us from regression on this bug?

We can revert all these lines, but unless we have tests, they can be later modified and the bug will not be noticed.

Thanks!

Reply:
Sorry, I was unclear: this line is covered by many tests, but they could not catch the bug introduced by `ensureBytes`.  

[`ensureBytes` returns the object unmodified if it is not a "unicode" object](https://github.com/twisted/twisted/pull/961/files#diff-d1da2d643fedb71fd3c50a44b847b30dR30):

```
>>> from twisted.python.bytes import ensureBytes
>>> type(ensureBytes(object()))
<type 'object'>
```

`dirObj` should _never_ be a string-like object, but `ensureBytes` allows its type to be either "unicode" or _anything_.  Imagine what would happen if you used `ensureBytes` in a buggy protocol implementation; it could erroneously pass an arbitrary object through to its transport (which could in turn be another protocol), eventually resulting in a type error deep inside Twisted.  Conversion functions like `ensureBytes` must always return objects of their claimed return type to callers to avoid propagating bugs.  

That's exactly the precaution taken by `networkString`.
[On Python 2](https://github.com/twisted/twisted/blob/trunk/src/twisted/python/compat.py#L535-L540)
```
>>> from twisted.python.compat import networkString
>>> networkString(object())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/mrw/src/python/twisted/src/twisted/python/compat.py", line 537, in networkString
    raise TypeError("Can only pass-through bytes on Python 2")
TypeError: Can only pass-through bytes on Python 2
```

[On Python 3](https://github.com/twisted/twisted/blob/trunk/src/twisted/python/compat.py#L521-L524)
```python
>>> from twisted.python.compat import networkString
>>> networkString(object())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/mrw/src/python/twisted/src/twisted/python/compat.py", line 523, in networkString
    raise TypeError("Can only convert text to bytes on Python 3")
TypeError: Can only convert text to bytes on Python 3
````

`ensureBytes` could be modified to check that it's received either a "unicode" or `bytes` object, encoding the former to bytes with UTF-8 or the provided codec, passing through the latter, and raising an exception otherwise.  But now then it would be mimicking `networkString`, except that it would allow callers to accept _both_ bytes and text; I can't think of any APIs that actually work this way, other than problematic ones like Python 3's file system path interface.  That's why this function is a net negative.

Reply:
I am happy to revert this, but I am trying to see if we can do more than just reverting the change and allowing similar bugs to surface in the future.

> Sorry, I was unclear: this line is covered by many tests, but they could not catch the bug introduced by ensureBytes.

A test just touching this line is not enough :)
I was asking if there is a test which can trigger this bug and which can fail if the objects are not right.

Since the tests can't catch the bugs, I would say that we have not tests :) .

I would like to use to opportunity to detect the place where encoding errors can occur and write explicit tests for this behaviour.

Now, we can merge this revert as it is and try to write more tests in a follow up... which might be never